File: 151.reverse-words-in-a-string.cpp
/*
 * @lc app=leetcode id=151 lang=cpp
 *
 * [151] Reverse Words in a String
 */

// @lc code=start
#include <sstream>
#include <stack>
#include <string>
using namespace std;

class Solution {
 public:
  string reverseWords(string s) {
    int j = s.size();
    string result = "";
    for (int i = s.size() - 1; i >= 0; i--) {
      if (s[i] == ' ') {
        j = i;
      } else if (i == 0 || s[i - 1] == ' ') {
        if (result.size() == 0) {
          result = s.substr(i, j - i);
        } else {
          result += " " + s.substr(i, j - i);
        }
      }
    }
    return result;
  }
};
// @lc code=end

File: 151.reverse-words-in-a-string.java
/*
 * @lc app=leetcode id=151 lang=java
 *
 * [151] Reverse Words in a String
 */

// @lc code=start
class Solution {
  public String reverseWords(String s) {
    int j = s.length();
    StringBuilder sb = new StringBuilder();
    for (int i = s.length() - 1; i >= 0; i--) {
      if (s.charAt(i) == ' ') {
        j = i;
      } else if (i == 0 || s.charAt(i - 1) == ' ') {
        if (sb.length() == 0) {
          sb.append(s.substring(i, j));
        } else {
          sb.append(' ' + s.substring(i, j));
        }
      }
    }
    return sb.toString();
  }
}

// @lc code=end
